packages/voting/src/deposit.rs

```diff
@@ -0,0 +1,133 @@
+    /// The address of the cw20 token to be used for proposal
+    /// deposits.
+    pub token: Addr,
+    /// The number of tokens that must be deposited to create a
+    /// proposal.
+    pub deposit: Uint128,
+    /// If failed proposals should have their deposits refunded.
+    pub refund_failed_proposals: bool,
+}
+
+impl DepositInfo {
+    /// Converts deposit info into checked deposit info.
+    pub fn into_checked(self, deps: Deps, dao: Addr) -> StdResult<CheckedDepositInfo> {
+        let Self {
+            token,
+            deposit,
+            refund_failed_proposals,
+        } = self;
+        let token = match token {
+            DepositToken::Token { address } => deps.api.addr_validate(&address)?,
+            DepositToken::VotingModuleToken {} => {
+                let voting_module: Addr = deps
+                    .querier
+                    .query_wasm_smart(dao, &cw_core::msg::QueryMsg::VotingModule {})?;
+                let token_addr: Addr = deps.querier.query_wasm_smart(
+                    voting_module,
+                    &cw_core_interface::voting::Query::TokenContract {},
+                )?;
+                token_addr
+            }
+        };
+        // Make an info query as a smoke test that we are indeed
+        // working with a token here. We can't turbofish this
+        // type. See <https://github.com/rust-lang/rust/issues/83701>
+        let _info: cw20::TokenInfoResponse = deps
+            .querier
+            .query_wasm_smart(token.clone(), &cw20::Cw20QueryMsg::TokenInfo {})?;
+        Ok(CheckedDepositInfo {
+            token,
+            deposit,
+            refund_failed_proposals,
+        })
+    }
+}
+
+pub fn get_deposit_msg(
+    info: &Option<CheckedDepositInfo>,
+    contract: &Addr,
+    sender: &Addr,
+) -> StdResult<Vec<CosmosMsg>> {
+    match info {
+        Some(info) => {
+            if info.deposit.is_zero() {
+                Ok(vec![])
+            } else {
+                let transfer_msg = WasmMsg::Execute {
+                    contract_addr: info.token.to_string(),
+                    funds: vec![],
+                    msg: to_binary(&cw20::Cw20ExecuteMsg::TransferFrom {
+                        owner: sender.to_string(),
+                        recipient: contract.to_string(),
+                        amount: info.deposit,
+                    })?,
+                };
+                let transfer_msg: CosmosMsg = transfer_msg.into();
+                Ok(vec![transfer_msg])
+            }
+        }
+        None => Ok(vec![]),
+    }
+}
+
+pub fn get_return_deposit_msg<T: Proposal>(proposal: &T) -> StdResult<Vec<CosmosMsg>> {
+    match &proposal.deposit_info() {
+        Some(info) => {
+            // If it is closed only issue a refund if we are refunding
+            // failed proposals. Otherwise, the proposal is open and
+            // expired.
+            if proposal.status() == Status::Rejected && !info.refund_failed_proposals {
```

Comment:
```suggestion
            if proposal.status() == Status::Rejected && !info.refund_failed_proposals || info.deposit.is_zero() {
```

Otherwise this will error with a zero proposal deposit. This is my bug, not yours. May as well fix it here. :)